### PR TITLE
Fixed bug when referencing variables from other variable object values

### LINKED ISSUE #3587: Validation Error Detected v1.13.0
As of the release of version 1.13.0 I get the following validation error message when running a deploy. This doesn't happen with version 1.12.1. I've only seen then when migrating directly from version 1.5.0 to 1.13.0. I haven't tested to see if it happens from any other migration paths. 
```
  An error occurred while provisioning your stack: media
     - 1 validation error detected: Value '${self:custom.databasePrefix}-media'
     at 'tableName' failed to satisfy constraint: Member
     must satisfy regular expression pattern: [a-zA-Z0-9_.-]+.
```

### PR DESCRIPTION
**Pull request by eahefnawy**
## What did you implement:

Closes #3587

Fixed a bug with the variable system where populating variables referenced from object values of other variables was broken.

## How did you implement it:
If the value of a variable is an object, we're now recursively populating that object. I've also added tests for this edge case.

## How can we verify it:
With the following two files:

serverless.yml
```yml
service: vars-in-files

provider:
  name: aws
  runtime: nodejs6.10

functions:
  hello:
    handler: handler.hello

resources:
  Outputs:
    varA:
      Value: ${file(./vars.json)}
      Export:
        Name: varA
```
vars.json
```json
{
  "vars": {
    "deep": "${self:provider.name}-something"
  }
}
```
run `sls package` and check the `varA` output. You should see the variable populated correctly as `aws-something`

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

File: lib/classes/Variables.test.js [hunk 1]
```diff
@@ -123,6 +157,44 @@ describe('Variables', () => {
...
+      return serverless.variables.populateProperty(property).then(newProperty => {
+        expect(populateObjectStub.called).to.equal(true);
+        expect(getValueFromSourceStub.called).to.equal(true);
+        expect(populateVariableStub.called).to.equal(true);
+        expect(newProperty).to.deep.equal(variableValuePopulated);
+
+        serverless.variables.populateObject.restore();
+        serverless.variables.getValueFromSource.restore();
+        serverless.variables.populateVariable.restore();
+        return Promise.resolve();
```
Comment (pmuens) [inline lib/classes/Variables.test.js]: Could we stick to one Promise implementation from now on (even in tests)?

I'll fix this one in the Async follow-up PR

**Comment by antanasbrazenas** [comment]
Introduction of async variable support started causing JavaScript heap out of memory issues for me.

**Comment by pmuens** [comment]
Thanks for the feedback @antanasbrazenas 👍 

We have a problem with an internal recursion causing an infinite loop. The problem is discussed and tracked in https://github.com/serverless/serverless/issues/3627.
